### ISSUE TITLE
Absent semigroup should short-circut folds on first nothing

### ISSUE THREAD

**Issue by jnape** [palatable/lambda#115]
`Absent#foldLeft` and `#foldRight` should process the minimal number of elements of the inbound `Iterable` necessary to produce the result, and since the semigroup treats `nothing` as an annihilator, that means that `nothing()` should cause a short-circuit. As a good heuristic, both `Absent.<Unit>absent(Constantly::constantly).foldLeft(nothing(), repeat(just(UNIT)));` and `Absent.<Unit>absent(Constantly::constantly).foldRight(nothing(), repeat(just(UNIT))).value();` should terminate.

**Comment by corlaez** [comment]
Am I going to need a new Semigroup class? AbsentSemigroup?

**Comment by jnape** [comment]
Hi @corlaez, welcome to the project and thank you for the kind words.

No, the current `Absent` should remain a `SemigroupFactory` with the same constraints, the only goal here is to short-circuit the folds. The way I'd approach it would be to replace line `35` with an anonymous instance of `Semigroup<Maybe<A>>` that overrides `foldLeft` and `foldRight` accounting for short-circuiting. Something like this:

```java
public final class Absent<A> implements SemigroupFactory<Semigroup<A>, Maybe<A>> {

    // ...

    @Override
    public Semigroup<Maybe<A>> checkedApply(Semigroup<A> aSemigroup) {
        return new Semigroup<Maybe<A>>() {
            @Override
            public Maybe<A> checkedApply(Maybe<A> maybeX, Maybe<A> maybeY) throws Throwable {
                return liftA2(aSemigroup, maybeX, maybeY);
            }

            @Override
            public Maybe<A> foldLeft(Maybe<A> aMaybe, Iterable<Maybe<A>> maybes) {
                throw new UnsupportedOperationException("implement me with short-circuiting");
            }

            @Override
            public Lazy<Maybe<A>> foldRight(Maybe<A> aMaybe, Iterable<Maybe<A>> maybes) {
                throw new UnsupportedOperationException("implement me with short-circuiting");
            }
        };
    }

    // ...
}
```

Does that make sense?

Also, please feel free to join the [lambda](https://gitter.im/palatable/lambda) gitter to chat with me directly, and again, welcome!

**Comment by corlaez** [comment]
That does make sense. Ok!
I was wondering if we needed to encode absorbing elements somehow but I think that will do.

**Comment by corlaez** [comment]
Making progress I have left fold working

**Comment by corlaez** [comment]
Struggling with fold Right... the typing and lazys got me a little confused
Is the answer going to be something like this? Should I avoid lazyRec?
```java
@Override
            public Lazy<Maybe<A>> foldRight(Maybe<A> acc, Iterable<Maybe<A>> as) {
                return lazyRec((f, lazyIt) -> ...
                                ? ...
                                : lazy(acc)),
                        lazy(as::iterator));
            }
```

**Comment by jnape** [comment]
@corlaez Some combination of `FoldRight.foldRight` and `lazyZip` ought to do it. Something like:

```java
FoldRight.foldRight(
        (maybeX, acc) -> maybeX.lazyZip(acc.fmap(maybeY -> maybeY.fmap(aSemigroup.flip()))),
        lazy(acc),
        rest);
```

Combine that with an initial `match` on the accumulator (so knowledge about a `nothing()` accumulator can be leveraged to elide the entire `foldRight`) and I think you're home.

Makes sense?

**Comment by corlaez** [comment]
PR open! Thanks for the guidance

**Comment by corlaez** [comment]
It leaves me wondering if what we want is a SemigroupWithZero class that can be tested independently and that can be used where it makes sense.
A SemigroupWithZero would define an absorbent element that can help to short-circuit folds in Absent and any other class that needs it.

**Comment by jnape** [comment]
Closed by #116
